index.coffee

```diff
@@ -1,60 +1,172 @@
-through = require 'through'
+stream = require 'readable-stream'
 
 module.exports = (requestModule, requestOpts, opts) ->
-	res = new Resumable(requestModule, requestOpts, opts)
-	return res.resume()
-
-class Resumable
-	constructor: (requestModule, requestOpts, opts) ->
-		@stream = through (data) =>
-			@bytesRead += data.length
-			@stream.emit('data', data)
+	return new ResumableRequest(requestModule, requestOpts, opts)
+
+module.exports.defaults = (defaults = {}) ->
+	(requestModule, requestOpts, opts) ->
+		opts = Object.assign {}, defaults, opts
+		return new ResumableRequest(requestModule, requestOpts, opts)
+
+cloneResponse = (response) ->
+	# creates an `http.IncomingMessage`-like stream, that can also be piped to.
+	s = new stream.PassThrough()
+	[ 'headers', 'httpVersion', 'method', 'rawHeaders', 'statusCode', 'statusMessage' ].forEach (prop) ->
+		s[prop] = response[prop]
+	return s
+
+throttle = (fn, interval) ->
+	timeoutId = null
+	cancel = ->
+		return if not timeoutId?
+		timeoutId = clearTimeout(timeoutId)
+	apply = ->
+		cancel()
+		fn()
+	throttledFn = ->
+		return if timeoutId?
+		timeoutId = setTimeout(apply, interval)
+	throttledFn.apply = apply
+	throttledFn.cancel = cancel
+	return throttledFn
+
+wrapLegacyStream = (s) ->
+	new stream.Readable().wrap(s)
+
+class ResumableRequest extends stream.Readable
+	constructor: (@requestModule, @requestOpts, opts = {}) ->
+		super()
+
+		@error = null
+		@request = null
+		@response = null
+
 		@bytesRead = 0
+		@bytesTotal = null
 		@retries = 0
-		@maxRetries = opts?.maxRetries ? 10
-		@contentLength = null
-		@requestOpts = requestOpts
-		@requestModule = requestModule
-		@error = null
+		@maxRetries = opts.maxRetries ? 10
+
+		@_retry = throttle(@_retry.bind(this), opts.retryInterval ? 1000)
+		@_reportProgress = throttle(@_reportProgress.bind(this), opts.progressInterval ? 1000)
+
+		@on 'pipe', ->
+			# request can be written to when uploading a resource to a URL.
+			# We do not support that -- resumable request can only be used
+			# to *download* a resource, i.e. only the response is resumable.
+			# piping will fail as soon as the stream gets going because we don't
+			# implement Writable but it's better to fail early.
+			throw new Error('ResumableRequest is not writable.')
```

Comment:
It will, and I think it's the correct thing to do -- this is programmer error and it shouldn't be discovered at runtime.